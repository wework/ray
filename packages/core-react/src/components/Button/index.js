import React from 'react';
import clsx from 'clsx';
import PropTypes from 'prop-types';

const Button = React.forwardRef((props, ref) => {
  const {
    Tag,
    children,
    variant,
    compact,
    danger,
    disabled,
    className,
    ...rest
  } = props;
  return (
    <Tag
      {...rest}
      className={clsx(
        'ray-button',
        `ray-button--${variant}`,
        {
          'ray-button--compact': compact,
          'ray-button--danger': danger
        },
        className
      )}
      disabled={disabled}
      ref={ref}
    >
      {children}
    </Tag>
  );
});

Button.propTypes = {
  variant: PropTypes.oneOf(['primary', 'secondary', 'tertiary']),
  compact: PropTypes.bool,
  danger: PropTypes.bool,
  children: PropTypes.node,
  Tag: PropTypes.oneOf(['button', 'a']),
  disabled: PropTypes.bool,
  className: PropTypes.string
};

Button.defaultProps = {
  Tag: 'button',
  variant: 'primary',
  compact: false,
  danger: false
};

Button.displayName = 'Button';

export default Button;
