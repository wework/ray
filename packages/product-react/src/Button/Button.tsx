import classNames from 'classnames/bind';
import React from 'react';

export interface ButtonProps extends React.HTMLProps<HTMLButtonElement> {
  compact?: boolean;
  danger?: boolean;
  tabIndex?: number;
  type?: 'primary' | 'secondary' | 'tertiary';
}

const styles = {
  primary: 'ray-button--primary',
  secondary: 'ray-button--secondary',
  tertiary: 'ray-button--tertiary',
  compact: 'ray-button--compact',
  danger: 'ray-button--danger',
};

const cx = classNames.bind(styles);

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    const {
      className,
      compact,
      danger,
      disabled,
      tabIndex,
      type,
      ...rest
    } = props;

    return (
      <button
        ref={ref}
        className={cx('ray-button', type, { compact, danger }, className)}
        disabled={disabled}
        tabIndex={tabIndex}
        {...rest}
      />
    );
  },
);

Button.defaultProps = {
  compact: false,
  danger: false,
  disabled: false,
  type: 'primary',
};

export default Button;
