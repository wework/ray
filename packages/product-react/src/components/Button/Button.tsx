import clsx from 'clsx';
import React, { createRef, forwardRef, ReactHTMLElement } from 'react';

export type ButtonProps = {
  children?: React.ReactNode;
  buttonStyle?: 'primary' | 'secondary' | 'tertiary';
  type?: 'button' | 'reset' | 'submit';
  compact?: boolean;
  danger?: boolean;
  disabled?: boolean;
  className?: string;
  tabIndex?: number;
  name?: string;
  value?: string | string[] | number;
  onClick?(event: React.MouseEvent<HTMLButtonElement>): void;
};

export const Button = React.forwardRef(
  (props: ButtonProps, ref: React.Ref<HTMLButtonElement>) => {
    const {
      name,
      value,
      children,
      type,
      compact,
      danger,
      disabled,
      className,
      buttonStyle,
      ...rest
    } = props;

    return (
      <button
        name={name}
        value={value}
        type={type}
        {...rest}
        className={clsx(
          'ray-button',
          `ray-button--${buttonStyle}`,
          {
            'ray-button--compact': compact,
            'ray-button--danger': danger
          },
          className
        )}
        disabled={disabled}
        ref={ref}
        {...rest}
      >
        {children}
      </button>
    );
  }
);

Button.defaultProps = {
  buttonStyle: 'primary',
  compact: false,
  danger: false,
  disabled: false
};

Button.displayName = 'Button';

export default Button;
