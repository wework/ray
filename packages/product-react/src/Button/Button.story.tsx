import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';
import React from 'react';

import { Button } from './Button';

storiesOf('Core Comps | Button', module)
  .add('Primary', () => {
    const onClick = action('clicked');
    return <Button>Primary Button</Button>;
  })
  .add('Primary w/ Icon', () => {
    const onClick = action('clicked');
    return (
      <Button onClick={onClick}>
        Primary Button w/ Icon
        <svg
          className="ray-button__icon"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 25 25"
        >
          <g>
            <path d="M12.5,6C5.5964,6,0,12.6211,0,13.5S5.5964,21,12.5,21,25,14.4375,25,13.5,19.4036,6,12.5,6Zm0,11A3.5,3.5,0,1,1,16,13.5,3.5,3.5,0,0,1,12.5,17Z" />
          </g>
        </svg>
      </Button>
    );
  })
  .add('Primary - Disabled', () => {
    const onClick = action('clicked');
    return (
      <Button onClick={onClick} disabled={true}>
        Primary Disabled Button
      </Button>
    );
  })
  .add('Secondary', () => {
    const onClick = action('clicked');
    return (
      <Button onClick={onClick} type="secondary">
        Secondary Button
      </Button>
    );
  })
  .add('Secondary - Disabled', () => {
    const onClick = action('clicked');
    return (
      <Button onClick={onClick} type="secondary" disabled={true}>
        Seconary Disabled Button
      </Button>
    );
  })
  .add('Tertiary', () => {
    const onClick = action('clicked');
    return (
      <Button onClick={onClick} type="tertiary">
        Tertiary Button
      </Button>
    );
  })
  .add('Tertiary - Disabled', () => {
    const onClick = action('clicked');
    return (
      <Button onClick={onClick} type="tertiary" disabled={true}>
        Tertiary Disabled Button
      </Button>
    );
  })
  .add('Compact', () => {
    const onClick = action('clicked');
    return (
      <Button onClick={onClick} compact={true}>
        Primary Compact Button
      </Button>
    );
  })
  .add('Compact - Disabled', () => {
    const onClick = action('clicked');
    return (
      <Button onClick={onClick} compact={true} disabled={true}>
        Primary Compact Disabled Button
      </Button>
    );
  })
  .add('Danger', () => {
    const onClick = action('clicked');
    return (
      <Button onClick={onClick} danger={true}>
        Primary Danger Button
      </Button>
    );
  })
  .add('Danger - Disabled', () => {
    const onClick = action('clicked');
    return (
      <Button onClick={onClick} danger={true} disabled={true}>
        Primary Danger Disabled Button
      </Button>
    );
  });
