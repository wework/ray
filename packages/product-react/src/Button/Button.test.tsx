import { mount, ReactWrapper } from 'enzyme';
import React from 'react';
import { Button } from './Button';

describe('Button', () => {
  let component: ReactWrapper<any>;
  const fauxCallback = jest.fn();

  beforeEach(() => {
    component = mount(<Button onClick={fauxCallback} />);
  });

  test('should match snapshot and styles for default props', () => {
    expect(component.debug()).toMatchSnapshot();
  });

  describe('Default Properties', () => {
    it('has default "TYPE" property that can be set', () => {
      expect(component.props().type).toEqual('primary');
      component.setProps({ type: 'secondary' });
      expect(component.props().type).toEqual('secondary');
      component.setProps({ type: 'tertiary' });
      expect(component.props().type).toEqual('tertiary');
    });

    it('has default "COMPACT" property that can be set', () => {
      expect(component.props().compact).toEqual(false);
      component.setProps({ compact: true });
      expect(component.props().compact).toEqual(true);
    });

    it('has default "DANGER" property that can be set', () => {
      expect(component.props().danger).toEqual(false);
      component.setProps({ danger: true });
      expect(component.props().danger).toEqual(true);
    });

    it('has default "DISABLED" property that can be set', () => {
      expect(component.props().disabled).toEqual(false);
      component.setProps({ disabled: true });
      expect(component.props().disabled).toEqual(true);
    });
  });

  describe('Classes', () => {
    it('contains DEFAULT', () => {
      expect(component.find('button').hasClass('ray-button')).toEqual(true);
    });

    it('contains correct class that corresponds with a specified TYPE prop', () => {
      expect(component.find('button').hasClass('ray-button--primary')).toEqual(
        true,
      );
      component.setProps({ type: 'secondary' });
      expect(
        component.find('button').hasClass('ray-button--secondary'),
      ).toEqual(true);
      component.setProps({ type: 'tertiary' });
      expect(component.find('button').hasClass('ray-button--tertiary')).toEqual(
        true,
      );
    });
  });

  describe('Composed Classes', () => {
    it('contains composed classes when it is COMPACT', () => {
      component.setProps({ compact: true });
      expect(
        component
          .find('button')
          .hasClass('ray-button--primary ray-button--compact'),
      ).toEqual(true);
      component.setProps({ type: 'secondary', compact: true });
      expect(
        component
          .find('button')
          .hasClass('ray-button--secondary ray-button--compact'),
      ).toEqual(true);
      component.setProps({ type: 'tertiary', compact: true });
      expect(
        component
          .find('button')
          .hasClass('ray-button--tertiary ray-button--compact'),
      ).toEqual(true);
    });
  });

  describe('Functions', () => {
    it('clicks', () => {
      component.find('button').simulate('click');
      expect(fauxCallback.mock.calls.length).toEqual(1);
    });
  });
});
