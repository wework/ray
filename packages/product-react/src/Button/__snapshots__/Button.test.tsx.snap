// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot and styles for default props 1`] = `
"<ForwardRef onClick={[Function: mockConstructor]} compact={false} danger={false} disabled={false} type=\\"primary\\">
  <button className=\\"ray-button ray-button--primary\\" disabled={false} tabIndex={[undefined]} onClick={[Function: mockConstructor]} />
</ForwardRef>"
`;
